keps/sig-storage/20191012-storagepool-advplacement.md

```diff
@@ -0,0 +1,364 @@
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story2-LocalDrives](./kep-storagepool-story2.png)
+
+Beyond the existence of the Placement Choice / Storage Pool, Minio further uses metadata to know that the storage pool has enough capacity to accommodate the desired PVC. If for example physical disks have different size, then Minio may decide to create PVCs of different sizes to still work in this scenario.
+
+#### Story 3
+
+Let’s assume a sharded and replicated MongoDB instance. Let’s further assume that the Cloud Storage Service used for the PVCs is comprised of 3 distinct Storage-Pods (not to be confused with K8s Pods) in both of the two available AZs, such that the 3 Storage-Pods fail independently by having dedicated power, networking, racks, hosts and disks. The CSI driver of the Cloud Storage service announces that 3 Storage-Pods per AZ as placement choices (AZ1-SP1, AZ1-SP2, AZ1-SP3, AZ2-SP1, etc.) and also advertises that all nodes in AZ1 can access all Storage-Pods in AZ1 (as a simple example). Let’s further assume MongoDB is configured for 3-way replication per shard. MongoDB decides to create 3 Pods for a given shard:
+* Pod1 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP1.
+* Pod2 has an affinity to AZ1, and its PVCs have an affinity to AZ1-SP2.
+* Pod3 has an affinity to AZ2, and its PVCs have an affinity to AZ2-SP1.
+
+If we compare this to the case where we can’t express the PVC affinity to the Storage-Pod, Pod1 and Pod2 may land on the same Storage-Pod and so if that Storage-Pod fails, it would take down two Pods. In the scenario described here, the math will show a higher availability and durability.
+
+The placement choices in this use case cannot be expressed using existing Kubernetes mechanism.
+
+![Story3-MultiStoragePods](./kep-storagepool-story3.png)
+
+
+### Implementation Details/Notes/Constraints
+
+#### New CRD to surface Storage Pools
+```
+apiVersion: storagepool.storage.k8s.io/v1alpha1
+kind: StoragePool
+metadata:
+  name: storagePool1
+  labels:
+     labelKey: labelValue
+spec:
+  driver: myDriver
+  parameters:
+     csiBackendSpecificOpaqueKey: csiBackendSpecificOpaqueValue
+status:
+  accessibleNodes: 
```

Comment:
Have we considered label selector for nodes instead of list of names? Maybe allow both (not together though)?

Reply:
Oh, somehow I managed to miss this field during my first pass through this KEP. I agree with @andrewsykim, relying exclusively on a list of node names is not going to scale for large clusters for pools which are accessible by all nodes.

In https://github.com/kubernetes/enhancements/pull/1353 I propose to use a `v1.NodeSelector` exactly for this reason.

Reply:
`accessibleNodes` here is under `status` so it should be a list of node names as the driver should know what nodes this pool is accessible to.

We could have a node selector in `spec`.

Reply:
Ah good point about status. I don't see `accessibleNodes` in spec, is the consumer of StoragePool API supposed to know that without input from the storage pool API?

Reply:
All, I don't quite follow the discussion. StoragePool is generated by code, in fact coming from CSI information. How would the CSI driver construct a NodeSelector, it doesn't really know the node labels, right? 

But why do we need the node selector in the spec? What would it express? Similar question for Andrew. My expectation is that the consumer (an application) would read the status of the storage pool and hence know the nodes.

Reply:
You are probably thinking of pools which only have a very short list of nodes. That may be valid for the intended purpose of this KEP.

But I am looking at it from the perspective of https://github.com/kubernetes/enhancements/pull/1353 where a StoragePool might be accessible from one half of a very large cluster. Then listing all nodes is not going to be very space-efficient. I also don't know how often nodes are added or removed - such changes might then force updating the StoragePool.

I think the ideal API should support both: NodeSelector and lists of node names. Then whoever creates those objects can choose the method that is more suitable.

Reply:
> StoragePool is generated by code, in fact coming from CSI information. How would the CSI driver construct a NodeSelector, it doesn't really know the node labels, right?

The CSI driver also doesn't know what the nodes are called in Kubernetes. So whoever calls the CSI driver will have to translate between the naming used by the CSI driver and the naming used by Kubernetes.

Reply:
OK, so you are suggesting that the controller may do something like add a label to all the nodes in "row 1" and then set up a NodeSelector accordingly. This way, the size of the StoragePool and any Pod that tries to affinititze to the nodes doesn't have a huge list to work off of. I think that makes sense. I will update the proposal accordingly.